### PR TITLE
Transfer song

### PR DESCRIPTION
**Pull request by theJenix**
I'm still fighting a bug, but there's a ton here to review, so I wanted to get some eyes on it while I fight this last issue.

### CODE REVIEW

**Comment by dgreenhalgh** [comment]
When you said ton to review, you weren't lying

**Comment by theJenix** [comment]
:laughing:

File: SoundStream-test/src/com/lastcrusade/soundstream/net/message/RequestSongMessageTest.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+import java.io.IOException;
+
+import org.junit.Test;
+
+public class RequestSongMessageTest extends SerializationTest<RequestSongMessage> {
+    
+    @Test
+    public void testSerializePlayMessage() throws IOException {
+        long songId = 8675309L;
```
Comment (dgreenhalgh) [inline SoundStream-test/src/com/lastcrusade/soundstream/net/message/RequestSongMessageTest.java:13]: :guitar:

File: SoundStream-test/src/com/lastcrusade/soundstream/net/message/TransferSongMessageTest.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import java.io.IOException;
+import java.util.Arrays;
+
+import org.junit.Test;
+
+public class TransferSongMessageTest extends SerializationTest<TransferSongMessage> {
+    
+    @Test
+    public void testSerializePlayMessage() throws IOException {
+        File file = new File("./assets/Jesse_normal_trimmed.wav");
```
Comment (dgreenhalgh) [inline SoundStream-test/src/com/lastcrusade/soundstream/net/message/TransferSongMessageTest.java:17]: Does this test only work on your device, @theJenix?

File: SoundStream/res/values/colors.xml [hunk 3]
```diff
@@ -5,6 +5,7 @@
     <color name="cornflower_blue">#6495ed</color>
     <color name="gray">#ffcccccc</color>
     <color name="transparent">#00000000</color>
+    <color name="loading">#22000000</color>
```
Comment (dgreenhalgh) [inline SoundStream/res/values/colors.xml:8]: Not sure what color loading is

File: SoundStream/src/com/lastcrusade/soundstream/model/Playlist.java [hunk 4]
```diff
@@ -1,60 +1,81 @@
...
+ *
+ */
 public class Playlist {
+    
     private final static String TAG = Playlist.class.getName();
-    private List<SongMetadata> musicList;
-    private int index;
+    
+    private Queue<PlaylistEntry> playedList;
+    private Queue<PlaylistEntry> musicList;
```
Comment (dgreenhalgh) [inline SoundStream/src/com/lastcrusade/soundstream/model/Playlist.java:20]: We need to make sure to pay attention to the distinction between these two Queues

File: SoundStream-test/src/com/lastcrusade/soundstream/net/message/TransferSongMessageTest.java (hunk 2 above)
Comment (reidbaker) [inline SoundStream-test/src/com/lastcrusade/soundstream/net/message/TransferSongMessageTest.java:17]: I think this will work because he included the file in the commit. :thumbsup:

File: SoundStream-test/src/com/lastcrusade/soundstream/net/message/TransferSongMessageTest.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        FileInputStream fis = new FileInputStream(file);
+        byte[] bytes = new byte[fis.available()];
+        fis.read(bytes);
+        TransferSongMessage oldMessage = new TransferSongMessage(132452L, file.getName(), bytes);
```
Comment (reidbaker) [inline SoundStream-test/src/com/lastcrusade/soundstream/net/message/TransferSongMessageTest.java]: Hard coded number

File: SoundStream-test/src/com/lastcrusade/soundstream/net/message/TransferSongMessageTest.java (hunk 2 above)
Comment (dgreenhalgh) [inline SoundStream-test/src/com/lastcrusade/soundstream/net/message/TransferSongMessageTest.java:17]: My bad, hadn't gotten that far when I made the comment

File: SoundStream/res/values/colors.xml (hunk 3 above)
Comment (theJenix) [inline SoundStream/res/values/colors.xml:8]: This is a light grey, to indicate that a file cannot actually be played yet...I intended for this to be a placeholder, until @ejohnson44 can style "loading" songs in a much better way.

File: SoundStream-test/src/com/lastcrusade/soundstream/net/message/TransferSongMessageTest.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        TransferSongMessage newMessage = super.testSerializeMessage(oldMessage);
+
+        assertEquals(oldMessage.getSongId(),       newMessage.getSongId());
```
Comment (reidbaker) [inline SoundStream-test/src/com/lastcrusade/soundstream/net/message/TransferSongMessageTest.java:26]: Seems to be some weird spacing here. Lets line all three of these up or reduce them to one space after the comma.

File: SoundStream-test/src/com/lastcrusade/soundstream/net/message/TransferSongMessageTest.java (hunk 5 above)
Comment (theJenix) [inline SoundStream-test/src/com/lastcrusade/soundstream/net/message/TransferSongMessageTest.java]: I'm not worried about these, because the test code needed some random number as the size, to ensure it gets serialized.  Therefore, the actual number is inconsequential...all that matters is that newMessage matches oldMessage.

File: SoundStream-test/src/com/lastcrusade/soundstream/service/MusicLibraryServiceTest.java [hunk 7]
```diff
@@ -74,7 +74,7 @@ public void testLibraryIsUnmodifiable() {
         // add one song, make sure we have 1 good item
```
Comment (reidbaker) [inline SoundStream-test/src/com/lastcrusade/soundstream/service/MusicLibraryServiceTest.java:74]: Just out of curiosity where did your ID number come from?

File: SoundStream-test/src/com/lastcrusade/soundstream/util/CustomAssert.java [hunk 8]
```diff
@@ -9,14 +9,16 @@
     /**
      * Test to ensure SongMetadata objects are equal.
      * 
-     * @param meta
-     * @param libraryItem
+     * @param expected
```
Comment (reidbaker) [inline SoundStream-test/src/com/lastcrusade/soundstream/util/CustomAssert.java:12]: Good refactor. My bad. Although it shouldn't matter I think there are tests that pass these in backwards.

File: SoundStream/res/values/colors.xml (hunk 3 above)
Comment (reidbaker) [inline SoundStream/res/values/colors.xml:8]: Color for the UI when a song is being transferred is what I am thinking. If that is incorrect we should change the name.

File: SoundStream-test/src/com/lastcrusade/soundstream/service/MusicLibraryServiceTest.java (hunk 7 above)
Comment (theJenix) [inline SoundStream-test/src/com/lastcrusade/soundstream/service/MusicLibraryServiceTest.java:74]: IDs started at 1 per mac address, and sizes were the result of me smacking the number row on my keyboard :smiley:

File: SoundStream-test/src/com/lastcrusade/soundstream/util/CustomAssert.java (hunk 8 above)
Comment (theJenix) [inline SoundStream-test/src/com/lastcrusade/soundstream/util/CustomAssert.java:12]: This is the order that the built in assert functions use.  We should make a note to fix those tests in a later commit...when viewing the results in the JUnit runner, it can be confusing if we reverse expected and actual.

File: SoundStream-test/src/com/lastcrusade/soundstream/net/message/TransferSongMessageTest.java (hunk 5 above)
Comment (reidbaker) [inline SoundStream-test/src/com/lastcrusade/soundstream/net/message/TransferSongMessageTest.java]: Then could we make it `long size = 132452L`

File: SoundStream-test/src/com/lastcrusade/soundstream/util/CustomAssert.java (hunk 8 above)
Comment (reidbaker) [inline SoundStream-test/src/com/lastcrusade/soundstream/util/CustomAssert.java:12]: Made a trello card. This is not an issue for this branch.

File: SoundStream/src/com/lastcrusade/soundstream/model/UserList.java [hunk 9]
```diff
@@ -63,6 +63,12 @@ public void removeUser(String macAddress){
         }
     }
 
+    public void copyFrom(UserList userList) {
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/model/UserList.java:66]: Interesting method I see where it is used. What advantage does this have over something like setUserList

File: SoundStream/src/com/lastcrusade/soundstream/components/PlaylistFragment.java [hunk 10]
```diff
@@ -42,14 +43,13 @@ public void onCreate(Bundle savedInstanceState){
                 this.getActivity(), PlaylistService.class, PlaylistService.PlaylistServiceBinder.class);
 
         final CustomApp curApp = (CustomApp) this.getActivity().getApplication();
-        mPlayListAdapter = new PlayListAdapter(this.getActivity(), new ArrayList<SongMetadata>(), curApp.getUserList());
+        mPlayListAdapter = new PlayListAdapter(this.getActivity(), Collections.EMPTY_LIST, curApp.getUserList());
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/components/PlaylistFragment.java:46]: Why this change? I recognize that this is probably better but if you don't ask you don't learn.

File: SoundStream-test/src/com/lastcrusade/soundstream/net/message/TransferSongMessageTest.java (hunk 5 above)
Comment (theJenix) [inline SoundStream-test/src/com/lastcrusade/soundstream/net/message/TransferSongMessageTest.java]: Oh damn I misread the code.  This isn't a size, this is an ID.  I'll pull it out into a variable.

File: SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+                        // keep track of the bytes requested, and the remote entries, so we
+                        // can avoid overloading the host, and we can clean up after ourselves
+                        loadRemote(entry);
+                        bytesRequested += entry.getFileSize();
+                        remotelyLoaded.add(entry);
+                    }
+                }
+                
+                try {
+                    Thread.sleep(1000);
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java]: Magic number.
Think this means try and load new songs every second, then try again.

Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java]: Good catch...I'll clarify

File: SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+
+    private void registerReceivers() {
+        this.registrar = new BroadcastRegistrar();
+        this.registrar
+            .addAction(MessagingService.ACTION_TRANSFER_SONG_MESSAGE, new IBroadcastActionHandler() {
+                
+                @Override
+                public void onReceiveAction(Context context, Intent intent) {
+                    String fromAddr = intent.getStringExtra(MessagingService.EXTRA_ADDRESS);
+                    long   songId   = intent.getLongExtra(  MessagingService.EXTRA_SONG_ID, -1 /*SongMetadata.UNKNOWN_SONG*/);
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java]: Not sure what is happening on this line

File: SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+        String compositeFileName = String.format("%s_%s", SongMetadataUtils.getUniqueKey(fromAddr, songId), fileName);
+        try {
+            FileOutputStream fos = this.application.openFileOutput(compositeFileName, Context.MODE_PRIVATE);
+            fos.write(fileData);
+            fos.close();
+            String filePath = this.application.getFileStreamPath(compositeFileName).getCanonicalPath();
+            entry.setFilePath(filePath);
+        } catch (IOException e) {
+            this.application.deleteFile(compositeFileName);
+            //TODO: set flag to indicate file is broken
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java:177]: Is this TODO for this pull request or another?

File: SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+        }
+        return found;
+    }
+
+    public void addToLoadQueue(PlaylistEntry entry) {
+        //if the entry isnt already loaded, add it to the load queue
+        //NOTE: if it is loaded, we can assume its already in remotelyLoaded
+        synchronized(entryMutex) {
+            if (!entry.isLoaded()) {
+//                if (this.remotelyLoaded.contains(entry)) {
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java:196]: Is this supposed to be commented.

File: SoundStream/src/com/lastcrusade/soundstream/components/PlaylistFragment.java (hunk 10 above)
Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/components/PlaylistFragment.java:46]: I had to change the ArrayList to use PlaylistEntry objects anyway, and I like the use of Collections.EMPTY_LIST because it's very obvious what we're doing there.

File: SoundStream/src/com/lastcrusade/soundstream/model/UserList.java (hunk 9 above)
Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/model/UserList.java:66]: We have a couple places in the code were we pass around the UserList, so we needed to be able to modify the contents of that object without replacing the object.  The same applied to Playlist, which is one of the reasons I changed PlaylistService to expose the entries, not the actual Playlist structure.

File: SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+        } catch (SongNotFoundException e) {
+            e.printStackTrace();
+        }
+    }
+    
+    private void loadRemote(PlaylistEntry entry) {
+        this.application.getMessagingService().sendRequestSongMessage(entry.getMacAddress(), entry.getId());
+    }
+
+    public void stopLoading() {
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java:225]: I think this is a place we should log what is happening. Based on my understanding of the code we should be call stop loading a bunch and when we do we should record it. Also magic number

File: SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java (hunk 14 above)
Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java:196]: I left it in because there's a bug I still need to fix here.

File: SoundStream/src/com/lastcrusade/soundstream/model/Playlist.java (hunk 4 above)
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/model/Playlist.java:20]: I thought we decided to move away from the 2 list implementation and have an index indicating where we were

File: SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java (hunk 13 above)
Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java:177]: I intended to take care of it in this pull request...what I may do, for ease of merging, is ask we pull it in as is, and I will fix this along with the Large File bug currently in trello.

File: SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java (hunk 12 above)
Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java]: It is getting a log valued extra from the intent, which requires that we specify a default value.  I'll make it clearer in code.

File: SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java (hunk 15 above)
Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java:225]: stopLoading gets called to kill the thread...it gets called when a phone is connected to a host, so we don't try and load things on the guest.  It should not get called often.

I'll clarify the magic number.

File: SoundStream/src/com/lastcrusade/soundstream/model/PlaylistEntry.java [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+        this.setAlbum(metadata.getAlbum());
+        this.setArtist(metadata.getArtist());
+        this.setId(metadata.getId());
+        this.setMacAddress(metadata.getMacAddress());
+        this.setFileSize(metadata.getFileSize());
+        this.setTitle(metadata.getTitle());
+    }
+
+    public boolean isLocalFile() {
+        //TODO: this should not reference bluetoothutils...instead, we should look at the user list or some
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/model/PlaylistEntry.java:31]: Is this TODO for this branch or another one?

File: SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java (hunk 15 above)
Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java:225]: In clarifying, I decided to up the number to 1500 ms.

File: SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java [hunk 17]
```diff
@@ -56,6 +64,39 @@ public MessageThread(BluetoothSocket socket, Handler handler, String disconnectA
...
+            @Override
+            public void run() {
+                try {
+                    while (mmWriteThreadRunning) {
+                        try {
+                            mmWriter.writeOne();
+                            Thread.sleep(10); //give the system a chance to breath
+                        } catch (IOException e) {
+                            //we've probably closed our socket...quit the thread
+                            mmWriteThreadRunning = false;
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java:80]: Do we need to inform the userlist the socket has gone down? Or some other piece of code?

File: SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java [hunk 18]
```diff
@@ -56,6 +64,39 @@ public MessageThread(BluetoothSocket socket, Handler handler, String disconnectA
...
+        mmWriteThreadRunning = true;
+        mmWriteThread = new Thread(new Runnable() {
+
+            @Override
+            public void run() {
+                try {
+                    while (mmWriteThreadRunning) {
+                        try {
+                            mmWriter.writeOne();
+                            Thread.sleep(10); //give the system a chance to breath
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java:77]: Do we need to stop here? If so what happens if we dont?

File: SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java (hunk 17 above)
Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java:80]: Probably not...this thread's execution is governed by MessageThread, which will also go down if the socket streams are throwing IOExceptions.  MessageThread provides more feedback, I believe.

File: SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java [hunk 19]
```diff
@@ -85,10 +126,26 @@ public void run() {
...
         //cancel and notify the handlers that the connection is dead
         cancel();
     }
  
+    private void stopWriteThread() {
+        mmStoppingThread = Thread.currentThread();
+        mmWriteThreadRunning = false;
+        synchronized(mmStoppingThread) {
+            try {
+                mmStoppingThread.wait(1000);
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java]: Magic number. Also I don't think I understand how this method is supposed to work. You may need to explain in person and that is cool.

File: SoundStream/src/com/lastcrusade/soundstream/model/PlaylistEntry.java (hunk 16 above)
Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/model/PlaylistEntry.java:31]: More of a long running TODO.  This was added as a note to the "Move user add out of CoreActivity" trello card.

File: SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java [hunk 20]
```diff
@@ -106,11 +163,13 @@ private void sendMessageToHandler(IMessage message, String remoteAddr) {
     }
 
     /* Call this from the main activity to send data to the remote device */
-    public void write(IMessage message) {
+    public synchronized void write(IMessage message) {
+        Log.d(TAG, "MessageThread#write called from " + Thread.currentThread().getName());
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java:170]: What is the # in there for?

Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java:170]: It's a convention for specifying member functions of a class.

File: SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java (hunk 19 above)
Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java]: I'll clarify, and yea we'll go over it when we meet tonight.

File: SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java (hunk 18 above)
Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java:77]: IOExceptions on socket based streams are usually the result of closing the socket.  As a result, if we don't stop, we'll encounter an unending number of IOExceptions.

File: SoundStream/src/com/lastcrusade/soundstream/net/message/ADataMessage.java [hunk 21]
```diff
@@ -82,19 +76,44 @@ protected long readLong(InputStream input) throws IOException {
...
+        }
+        else {
+            writeInteger(bytes.length, output);
+            output.write(bytes);
+        }
+    }
+
+    protected byte[] readBytes(InputStream input) throws IOException {
+        int length = readInteger(input);
+        //TODO: should put an upper bound here, and use a ByteArrayOutputStream to accumulate bytes
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/net/message/ADataMessage.java:109]: Is this TODO for this pull or another. If another lets make a trello card

File: SoundStream/src/com/lastcrusade/soundstream/service/ConnectionService.java [hunk 22]
```diff
@@ -355,6 +359,15 @@ public void broadcastMessageToGuests(IMessage msg) {
         }
     }
 
+    public void sendMessageToGuest(String address, IMessage msg) {
+        MessageThread fan = findMessageThreadByAddress(address);
+        if (fan != null) {
+            fan.write(msg);
+        } else {
+            Toaster.eToast(this, "Fan not connected");
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/service/ConnectionService.java:367]: Messages that the user sees should come from error_strings.xml

File: SoundStream/src/com/lastcrusade/soundstream/net/message/ADataMessage.java (hunk 21 above)
Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/net/message/ADataMessage.java:109]: Trello card added.

File: SoundStream/src/com/lastcrusade/soundstream/service/ConnectionService.java (hunk 22 above)
Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/service/ConnectionService.java:367]: I'm not so sure this is something a user should see.  I added it in for debug purposes.  This is more an exception than anything else.

File: SoundStream/src/com/lastcrusade/soundstream/service/MusicLibraryService.java [hunk 23]
```diff
@@ -216,33 +240,52 @@ void removeLibraryForAddress(String macAddress, boolean notify) {
...
      * Orders the song metadata and related map alphabetically by Artist,
      * Album, and Title
      */
     private void orderAlphabetically(){
-        synchronized (metadataMutex) {
+        synchronized(metadataMutex) {
             //sort the metadata alphabetically
             Collections.sort(metadataList, new AlphabeticalComparator());
             
             //recreate the map
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/service/MusicLibraryService.java:251]: Save this for another pull request. Ill make a trello card.

Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/service/MusicLibraryService.java:251]: I'm confused...what do we need to save for another pull request?

File: SoundStream/src/com/lastcrusade/soundstream/service/PlaylistService.java [hunk 24]
```diff
@@ -163,90 +179,140 @@ public void onReceiveAction(Context context, Intent intent) {
             @Override
             public void onReceiveAction(Context context, Intent intent) {
                 List<SongMetadata> newList = intent.getParcelableArrayListExtra(MessagingService.EXTRA_SONG_METADATA);
-                mPlaylist = new Playlist(newList);
+                mPlaylist.clear();
+                for (SongMetadata metadata : newList) {
+                    PlaylistEntry entry = new PlaylistEntry(metadata);
+                    //for the UI, assume that the song is "loaded" when displayed on the remote guest
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/service/PlaylistService.java:185]: Worth having a conversation about if the guests should care if the songs are loaded on the host.

Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/service/PlaylistService.java:185]: I think they should, because the playlist will play things "out of order" if a song is not loaded.  Lets discuss.

File: SoundStream/src/com/lastcrusade/soundstream/service/PlaylistService.java [hunk 25]
```diff
@@ -163,90 +179,140 @@ public void onReceiveAction(Context context, Intent intent) {
...
+            }
+            //still no available music..this means we're waiting for data to come in
+            //...display a warning, but don't play.
+            if (song == null) {
+                //TODO: instead of this, we may want to repost a message to wait for the next song to be available
+                //stop the player
+                this.mAudioPlayer.stop();
+                //pop up the notice
+                Toaster.iToast(this, getString(R.string.no_available_songs));
+            } else {
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/service/PlaylistService.java]: I would add a comment that this is the normal flow the above catch edge cases

File: SoundStream/src/com/lastcrusade/soundstream/util/MusicListAdapter.java [hunk 26]
```diff
@@ -16,16 +16,16 @@
 import com.lastcrusade.soundstream.model.User;
 import com.lastcrusade.soundstream.model.UserList;
 
-public class MusicListAdapter extends BaseAdapter {
+public class MusicListAdapter<T extends SongMetadata> extends BaseAdapter {
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/util/MusicListAdapter.java:19]: Why did you remove the SongMetadata's from each list? Im not sure T is any clearer.

File: SoundStream/src/com/lastcrusade/soundstream/util/SongMetadataUtils.java [hunk 27]
```diff
@@ -0,0 +1,22 @@
+package com.lastcrusade.soundstream.util;
+
+import com.lastcrusade.soundstream.model.SongMetadata;
+
+public class SongMetadataUtils {
+
+    /**
+     * Create a unique key for this song.  This unique key consists of:
+     *  Mac address (uniquely identifies a device)
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/util/SongMetadataUtils.java:9]: Since we are storing Mac address as a string we might want to warn that it may not be a valid mac.

File: SoundStream/src/com/lastcrusade/soundstream/util/SongMetadataUtils.java [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+     */
+    public static String getUniqueKey(SongMetadata song) {
+        return getUniqueKey(song.getMacAddress(), song.getId());
+    }
+    
+    public static String getUniqueKey(String songSourceAddress, long songId) {
+        return songSourceAddress.replace(":", "") + "_" + songId;
+    }
+}
\ No newline at end of file
```
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/util/SongMetadataUtils.java]: newline

File: SoundStream/src/com/lastcrusade/soundstream/util/MusicListAdapter.java (hunk 26 above)
Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/util/MusicListAdapter.java:19]: PlaylistEntry extends SongMetadata, and contains attributes that are needed to draw the playlist.  This is so we can use the MusicListAdapter with SongMetadata and derived classes.  See the list adapter in PlaylistFragment

File: SoundStream/src/com/lastcrusade/soundstream/util/SongMetadataUtils.java (hunk 28 above)
Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/util/SongMetadataUtils.java]: Fixed.

File: SoundStream/src/com/lastcrusade/soundstream/model/UserList.java (hunk 9 above)
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/model/UserList.java:66]: Cool I knew there was a reason.

File: SoundStream/src/com/lastcrusade/soundstream/util/SongMetadataUtils.java (hunk 27 above)
Comment (theJenix) [inline SoundStream/src/com/lastcrusade/soundstream/util/SongMetadataUtils.java:9]: I don't think people will make that mistake, because we never represent the output as a valid address.

File: SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java (hunk 13 above)
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/manager/PlaylistDataManager.java:177]: I think that is a solid plan.

File: SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java (hunk 18 above)
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java:77]: That seems like a good time to stop then.
"Stop :hammer: time"

File: SoundStream/src/com/lastcrusade/soundstream/service/MusicLibraryService.java (hunk 23 above)
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/service/MusicLibraryService.java:251]: Mistake should be on line 263 for the dont reference bluetooth utils

File: SoundStream/src/com/lastcrusade/soundstream/util/MusicListAdapter.java (hunk 26 above)
Comment (reidbaker) [inline SoundStream/src/com/lastcrusade/soundstream/util/MusicListAdapter.java:19]: Ok I think I understand. Sorry it was a bunch of state to keep in my head.

**Comment by reidbaker** [comment]
@theJenix I am at the point where I am ready to pull this in. It is huge but you have addressed almost all my comments.

**Comment by dgreenhalgh** [comment]
As we talk about pulling this in, what's our ETA for the big bug fix, @theJenix?

**Comment by theJenix** [comment]
Probably tomorrow, during our work session.  The current code should work for FB Pop, Join Hangout and Hangout.

**Comment by dgreenhalgh** [comment]
I thought you said we would have that before the presentation

**Comment by theJenix** [comment]
I tried, but I hit a bit of a wall with Android.  I fixed the issue where the sending phone was locking up, but to play a large song requires us to change how we're sending the TransferData message through the broadcast intent, in MessagingService.  It shouldn't take me too long to fix, but I don't feel good about trying to squeeze that in before the demo.

**Comment by dgreenhalgh** [comment]
Alright, that's fine. Let's just make sure to tailor the technical demo so that we safely avoid the bug :smile:

**Comment by reidbaker** [comment]
Can I do anything to help get this in before the presentation?
